libp2p-websocket/src/browser.rs

```diff
@@ -0,0 +1,281 @@
+            }
+
+            let _ = message_tx.unbounded_send(Err(IoError::new(IoErrorKind::ConnectionRefused,
+                                                  "close event on the websocket")));
+        };
+
+        js!{
+            var socket = @{websocket};
+            var open_cb = @{open_cb};
+            var message_cb = @{message_cb};
+            var close_cb = @{close_cb};
+            socket.addEventListener("open", function(event) {
+                open_cb();
+            });
+            socket.addEventListener("message", function(event) {
+                var reader = new FileReader();
+                reader.addEventListener("loadend", function() {
+                    var typed = new Uint8Array(reader.result);
+                    message_cb(typed);
+                });
+                reader.readAsArrayBuffer(event.data);
+            });
+            socket.addEventListener("close", function(event) {
+                close_cb();
+            });
+        };
+
+        Ok(open_rx.then(|result| {
+            match result {
+                Ok(Ok(r)) => Ok(r),
+                Ok(Err(e)) => Err(e),
+                // `Err` would happen here if `open_tx` is destroyed. `open_tx` is captured by
+                // the `WebSocket`, and the `WebSocket` is captured by `open_cb`, which is itself
+                // captured by the `WebSocket`. Due to this cyclic dependency, `open_tx` should
+                // never be destroyed.
+                // TODO: how do we break this cyclic dependency? difficult question
+                Err(_) => unreachable!("the sending side will only close when we drop the future")
+            }
+        }))
+    }
+}
+
+pub struct Connec {
+    websocket: Reference,
+    // Stream of messages that goes through a `RwStreamSink` in order to become a `AsyncRead`.
+    incoming_data: RwStreamSink<StreamThen<mpsc::UnboundedReceiver<Result<Vec<u8>, IoError>>,
+                                           fn(Result<Result<Vec<u8>, IoError>, ()>) -> Result<Vec<u8>, IoError>,
+                                           Result<Vec<u8>, IoError>
+                                           >>,
+}
+
+impl Drop for Connec {
+    #[inline]
+    fn drop(&mut self) {
+        // TODO: apparently there's a memory leak related to callbacks?
+        js!{ @{&self.websocket}.close(); }
+    }
+}
+
+impl AsyncRead for Connec {
+}
+
+impl Read for Connec {
+    #[inline]
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, IoError> {
+        self.incoming_data.read(buf)
+    }
+}
+
+impl AsyncWrite for Connec {
+    #[inline]
+    fn shutdown(&mut self) -> Poll<(), IoError> {
+        Ok(Async::Ready(()))
+    }
+}
+
+impl Write for Connec {
+    fn write(&mut self, buf: &[u8]) -> Result<usize, IoError> {
+        let typed_array = TypedArray::from(buf);
+
+        // `send` can throw if the websocket isn't open (which can happen if it was closed by the
+        // remote).
+        let returned = js!{
```

Comment:
Here and elsewhere: blocks should start with a space (so `js! { }` instead of `js!{ }`)